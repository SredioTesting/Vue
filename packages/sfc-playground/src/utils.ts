export function debounce(fn: Function, n = 100) {
  let handle: number | null = null
  return (...args: any[]) => {
    if (handle) window.clearTimeout(handle)
    handle = window.setTimeout(() => {
      fn(...args)
      handle = null
    }, n)
  }
}

// prefer old unicode hacks for backward compatibility
// https://base64.guru/developers/javascript/examples/unicode-strings
export function utoa(data: string): string {
  return btoa(unescape(encodeURIComponent(data)))
}

export function atou(base64: string): string {
  return decodeURIComponent(escape(atob(base64)))
}
