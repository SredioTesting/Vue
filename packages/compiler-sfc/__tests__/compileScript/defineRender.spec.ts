import { assertCode, compileSFCScript as compile } from '../utils'

describe('defineRender()', () => {
  test('JSX Element', () => {
    const { content } = compile(
      `
      <script setup lang="tsx">
      defineRender(<div />)
      </script>
    `,
      { defineRender: true },
    )
    assertCode(content)
    expect(content).toMatch(`return () => <div />`)
    expect(content).not.toMatch('defineRender')
  })

  test('function', () => {
    const { content } = compile(
      `
      <script setup lang="tsx">
      defineRender(() => <div />)
      </script>
    `,
      { defineRender: true },
    )
    assertCode(content)
    expect(content).toMatch(`return () => <div />`)
    expect(content).not.toMatch('defineRender')
  })

  test('identifier', () => {
    const { content } = compile(
      `
      <script setup lang="ts">
      import { renderFn } from './ctx'
      defineRender(renderFn)
      </script>
    `,
      { defineRender: true },
    )
    assertCode(content)
    expect(content).toMatch(`return renderFn`)
    expect(content).not.toMatch('defineRender')
  })

  test('empty argument', () => {
    const { content } = compile(
      `
      <script setup>
      const foo = 'bar'
      defineRender()
      </script>
    `,
      { defineRender: true },
    )
    assertCode(content)
    expect(content).toMatch(`return { foo }`)
    expect(content).not.toMatch('defineRender')
  })

  describe('errors', () => {
    test('w/ <template>', () => {
      expect(() =>
        compile(
          `
      <script setup lang="tsx">
      defineRender(<div />)
      </script>
      <template>
        <span>hello</span>
      </template>
    `,
          { defineRender: true },
        ),
      ).toThrow(`defineRender() cannot be used with <template>.`)
    })
  })
})
