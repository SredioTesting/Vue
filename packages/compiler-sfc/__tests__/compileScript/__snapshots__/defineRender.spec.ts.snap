// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`defineRender() > JSX Element 1`] = `
"import { defineComponent as _defineComponent } from 'vue'

export default /*#__PURE__*/_defineComponent({
  setup(__props, { expose: __expose }) {
  __expose();

      
      
return () => <div />
}

})"
`;

exports[`defineRender() > empty argument 1`] = `
"const foo = 'bar'
      
export default {
  setup(__props, { expose: __expose }) {
  __expose();

      
      
return { foo }
}

}"
`;

exports[`defineRender() > function 1`] = `
"import { defineComponent as _defineComponent } from 'vue'

export default /*#__PURE__*/_defineComponent({
  setup(__props, { expose: __expose }) {
  __expose();

      
      
return () => <div />
}

})"
`;

exports[`defineRender() > identifier 1`] = `
"import { defineComponent as _defineComponent } from 'vue'
import { renderFn } from './ctx'
      
export default /*#__PURE__*/_defineComponent({
  setup(__props, { expose: __expose }) {
  __expose();

      
      
return renderFn
}

})"
`;
